File: Data Structures & Algorithms/house-robber/submission-8.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)

        if n == 0:
            return 0
        elif n == 1:
            return nums[0]
        
        rob1, rob2 = nums[0], max(nums[0], nums[1])

        for i in range(2, n):
            rob1, rob2 = rob2, max(nums[i] + rob1, rob2)
        
        return rob2

File: Data Structures & Algorithms/house-robber-ii/submission-8.java
class Solution {
    public int robber(int[] nums) {
        int rob1 = 0, rob2 = 0;

        for (int n : nums) {
            int temp = rob2;
            rob2 = Math.max(n + rob1, rob2);
            rob1 = temp;
        }

        return rob2;
    }

    public int rob(int[] nums) {
        int temp = Math.max(nums[0], robber(Arrays.copyOfRange(nums, 1, nums.length)));

        return Math.max(temp, robber(Arrays.copyOfRange(nums, 0, nums.length - 1)));
    }
}
